autonomi/src/networking/driver/task_handler.rs

```diff
@@ -0,0 +1,325 @@
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use ant_evm::PaymentQuote;
+use ant_protocol::{NetworkAddress, PrettyPrintRecordKey};
+use libp2p::kad::{self, PeerInfo, QueryId, Record};
+use libp2p::request_response::OutboundRequestId;
+use libp2p::PeerId;
+use std::collections::HashMap;
+use thiserror::Error;
+
+use crate::networking::interface::NetworkTask;
+use crate::networking::NetworkError;
+use crate::networking::OneShotTaskResult;
+
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum TaskHandlerError {
+    #[error("No tasks matching query {0}, query might have been completed already")]
+    UnknownQuery(String),
+    #[error("Network client dropped, cannot send oneshot response")]
+    NetworkClientDropped,
+}
+
+type QuoteDataType = u32;
+type RecordAndHolders = (Option<Record>, Vec<PeerId>);
+
+/// The [`TaskHandler`] is responsible for handling the progress in pending tasks using the results from [`crate::driver::NetworkDriver::process_swarm_event`]
+/// Once a task is completed, the [`TaskHandler`] will send the result to the client [`crate::Network`] via the oneshot channel provided when the task was created
+///
+/// All fields in this struct are private so we know that only the code in this module can MUTATE them
+#[allow(clippy::type_complexity)]
+#[derive(Default)]
+pub(crate) struct TaskHandler {
+    closest_peers: HashMap<QueryId, OneShotTaskResult<Vec<PeerInfo>>>,
+    put_record: HashMap<QueryId, OneShotTaskResult<()>>,
+    get_cost: HashMap<
+        OutboundRequestId,
+        (
+            OneShotTaskResult<Option<(PeerInfo, PaymentQuote)>>,
+            QuoteDataType,
+            PeerInfo,
+        ),
+    >,
+    get_record: HashMap<QueryId, OneShotTaskResult<RecordAndHolders>>,
+    get_record_accumulator: HashMap<QueryId, HashMap<PeerId, Record>>,
+}
+
+impl TaskHandler {
+    pub fn contains(&self, id: &QueryId) -> bool {
+        self.closest_peers.contains_key(id)
+            || self.get_record.contains_key(id)
+            || self.put_record.contains_key(id)
+    }
+
+    pub fn contains_query(&self, id: &OutboundRequestId) -> bool {
+        self.get_cost.contains_key(id)
+    }
+
+    pub fn insert_task(&mut self, id: QueryId, task: NetworkTask) {
+        info!("New task: with QueryId({id}): {task:?}");
+        match task {
+            NetworkTask::GetClosestPeers { resp, .. } => {
+                self.closest_peers.insert(id, resp);
+            }
+            NetworkTask::GetRecord { resp, .. } => {
+                self.get_record.insert(id, resp);
+            }
+            NetworkTask::PutRecord { resp, .. } => {
+                self.put_record.insert(id, resp);
+            }
+            _ => {}
```

Comment:
`insert_query` shall not be separated from `insert_task`, also shall not use `wild card` here.

got changed in the PR https://github.com/maidsafe/autonomi/pull/2969

Reply:
Oh I didn't know, sure can fix that! What was the reason for that ban again?

Reply:
to avoid future new entry got slipped through in some hadling process